### PR TITLE
Shorten #file and add #filePath (behind an experimental flag)

### PR DESCRIPTION
**Pull request by beccadax**
`#file` includes an entire gigantic path whenever it is used, and it is often used implicitly. This has two negative effects: It wastes space on something that is almost never needed and it invisibly embeds information in the binary, such as the developer's username, that they may want to keep private.

This PR changes the behavior of `#file` so that it evaluates to a short string containing the module name and filename, but not the full path. It also adds `#filePath`, which continues to give you the absolute path.

These changes are hidden behind a feature flag, `-enable-experimental-concise-pound-file`, because making them permanent [will require an evolution proposal](https://forums.swift.org/t/pitch-concise-magic-file-names/31297).

This change includes some basic tests with the flag enabled, but they're not comprehensive; for instance, code completion isn't tested anywhere. The implementation for the proposal should update existing `#file` tests and add matching `#filePath` tests to ensure adequate coverage.

Progresses towards fixing rdar://problem/56816166.

### CODE REVIEW

**Comment by brentdax** [comment]
@swift-ci please benchmark

**Comment by brentdax** [comment]
Okay, "up to 6% code size" is probably a fair summary of that. Let's try compiler performance and look at number of bytes output as a rough proxy for code size of the source compatibility suite.

@swift-ci please test compiler performance

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
@swift-ci please check compiler performance

**Comment by brentdax** [comment]
@swift-ci please benchmark

**Comment by brentdax** [comment]
@swift-ci please test compiler performance

**Comment by brentdax** [comment]
macOS test failures are examples of the feature working correctly, but affecting tests I didn't update. I'm not sure why they didn't run locally. 🤷‍♂️

File: include/swift/Basic/LangOptions.h [hunk 1]
```diff
@@ -94,6 +94,10 @@ namespace swift {
     /// when using RequireExplicitAvailability.
     std::string RequireExplicitAvailabilityTarget;
 
+    /// If true, '#file' evaluates to the full path rather than a
+    /// human-readable string.
+    bool MagicFileIdentifierEvaluatesToPath = false;
```
Comment (brentdax) [inline include/swift/Basic/LangOptions.h]: Without an evolution proposal, this will need to be `true`.

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
The test failure is because I need to re-gyb SwiftSyntax files.

**Comment by brentdax** [comment]
apple/swift-syntax#185

@swift-ci please test

**Comment by brentdax** [comment]
apple/swift-syntax#185

@swift-ci please test

**Review by jckarter — APPROVED** [review]
LGTM.

**Comment by brentdax** [comment]
apple/swift-syntax#185

@swift-ci please clean test

**Pull request merged**